### PR TITLE
Do i2c chip detection before issuing any command

### PR DESCRIPTION
**Pull request by conchyliculture**

### CODE REVIEW

File: win/NvapiFansCli/NvapiFansCli.cpp [hunk 1]
```diff
@@ -164,6 +159,20 @@ int main(int argc, char* argv[])
 			std::cout << "Nvapi version:" << version << std::endl;
 		}
 
+		std::vector<NV_PHYSICAL_GPU_HANDLE> list_gpu = getAllGPUs(api);
+		bool detected = true;
+		for (NV_PHYSICAL_GPU_HANDLE& gpu : list_gpu) {
```
Comment (x0rbl) [inline win/NvapiFansCli/NvapiFansCli.cpp]: No reference needed

File: win/NvapiFansCli/NvapiFansCli.cpp [hunk 2]
```diff
@@ -47,17 +47,22 @@ bool showGPUInfos(NvApiClient api, NV_PHYSICAL_GPU_HANDLE gpu) {
 	return res;
 }
 
+std::vector<NV_PHYSICAL_GPU_HANDLE> getAllGPUs(NvApiClient api) {
```
Comment (x0rbl) [inline win/NvapiFansCli/NvapiFansCli.cpp]: `api` should be const reference

File: win/NvapiFansCli/NvapiFansCli.cpp [hunk 3]
```diff
@@ -66,7 +71,7 @@ bool showAllGPUsInfos(NvApiClient api, int gpuId) {
 	std::cout << "Found " << list_gpu.size() << " NVidia GPUs." << std::endl;
 
 	int index = 0;
-	for (NV_PHYSICAL_GPU_HANDLE gpu : list_gpu) {
+	for (NV_PHYSICAL_GPU_HANDLE &gpu : list_gpu) {
```
Comment (x0rbl) [inline win/NvapiFansCli/NvapiFansCli.cpp]: No reference needed (might be resolved with a merge?)

File: win/NvapiFansLib/NvapiFansLib.cpp [hunk 4]
```diff
@@ -250,3 +250,26 @@ int NvApiClient::getGPUTemperature(NV_PHYSICAL_GPU_HANDLE& handle) {
 	}
 	return -1;
 }
+
+bool NvApiClient::detectI2CDevice(NV_PHYSICAL_GPU_HANDLE& handle) {
+	bool res;
+	// Asus GPUTweakII detection mechanism
+	NvU8 high ;
```
Comment (x0rbl) [inline win/NvapiFansLib/NvapiFansLib.cpp]: Can remove space

File: win/NvapiFansSvc/NvapiFansSvc.cpp [hunk 5]
```diff
@@ -253,13 +253,22 @@ DWORD WINAPI ServiceWorkerThread(LPVOID lpParam)
     }
     LogSuccess(event_log, L"Config loaded");
 
+    std::vector<NV_PHYSICAL_GPU_HANDLE> list_gpu;
+    bool detected = true;
+    for (NV_PHYSICAL_GPU_HANDLE& gpu : list_gpu) {
```
Comment (x0rbl) [inline win/NvapiFansSvc/NvapiFansSvc.cpp]: No reference needed

File: win/NvapiFansCli/NvapiFansCli.cpp [hunk 6]
```diff
@@ -94,18 +99,13 @@ bool validateGPUId(std::vector<NV_PHYSICAL_GPU_HANDLE> list_gpu, int gpuId) {
```
Comment (x0rbl) [inline win/NvapiFansCli/NvapiFansCli.cpp]: Line 93: The `gpuId > 0 &&` part of the conditional is unnecessary:

`if (gpuId > 0 && gpuId >= (int)list_gpu.size()) {`

**Review by x0rbl — CHANGES_REQUESTED** [review]

File: win/NvapiFansCli/NvapiFansCli.cpp (hunk 1 above)
Comment (conchyliculture) [reply win/NvapiFansCli/NvapiFansCli.cpp]: done

File: win/NvapiFansCli/NvapiFansCli.cpp (hunk 2 above)
Comment (conchyliculture) [reply win/NvapiFansCli/NvapiFansCli.cpp]: done

File: win/NvapiFansCli/NvapiFansCli.cpp (hunk 3 above)
Comment (conchyliculture) [reply win/NvapiFansCli/NvapiFansCli.cpp]: yep

File: win/NvapiFansSvc/NvapiFansSvc.cpp (hunk 5 above)
Comment (conchyliculture) [reply win/NvapiFansSvc/NvapiFansSvc.cpp]: done

File: win/NvapiFansCli/NvapiFansCli.cpp (hunk 6 above)
Comment (conchyliculture) [reply win/NvapiFansCli/NvapiFansCli.cpp]: done

**Review by x0rbl — APPROVED** [review]

**Pull request merged**